tests/unit_tests/dft/include/reference_dft.hpp

```diff
@@ -23,44 +23,106 @@
 #include <cmath>
 #include <complex>
 #include <vector>
+#include <numeric>
 
+#include <oneapi/mkl/exceptions.hpp>
 #include "test_common.hpp"
 
-/** Naive DFT implementation for reference.
- *  
- * Directly compute a single 1D forward DFT of the form:
- * for k in range(0, N):
- *   out[k] = sum( exp(2 pi k n im / N) * in[n] for n in range(0, N) )
- * where N is the size of the input / output arrays. The input may be
- * real or complex, but the output must be complex. Unit strides are used
- * with no offset.
- * 
- * @tparam TypeIn The forward data type.
- * @tparam TypeOut The transformed (backward) data type. Written to. Must be 
- * complex.
- * @param in The input forward data.
- * @param out Where to write the output data.
-**/
+namespace detail {
 template <typename TypeIn, typename TypeOut>
-void reference_forward_dft(const std::vector<TypeIn> &in, std::vector<TypeOut> &out) {
-    if (in.size() != out.size()) {
-        throw std::invalid_argument("Input and output vectors must be of equal size.");
-    }
+void reference_forward_dft_impl(TypeIn *in, TypeOut *out, size_t N, size_t stride) {
     using ref_t = long double; /* Do the calculations using long double */
     static_assert(is_complex<TypeOut>(), "Output type of DFT must be complex");
 
-    const ref_t TWOPI = 2.0L * 3.141592653589793238462643383279502884197L;
+    constexpr ref_t TWOPI = 2.0L * 3.141592653589793238462643383279502884197L;
 
-    const size_t N = out.size();
     for (std::size_t k = 0; k < N; ++k) {
         std::complex<ref_t> out_temp = 0;
         const auto partial_expo = (static_cast<ref_t>(k) * TWOPI) / static_cast<ref_t>(N);
         for (std::size_t n = 0; n < N; ++n) {
             const auto expo = static_cast<ref_t>(n) * partial_expo;
-            out_temp += static_cast<std::complex<ref_t>>(in[n]) *
+            out_temp += static_cast<std::complex<ref_t>>(in[n * stride]) *
                         std::complex<ref_t>{ std::cos(expo), -std::sin(expo) };
         }
-        out[k] = static_cast<TypeOut>(out_temp);
+        out[k * stride] = static_cast<TypeOut>(out_temp);
+    }
+}
+
+template <typename TypeIn, typename TypeOut, int dims>
+struct reference {};
+
+template <typename TypeIn, typename TypeOut>
+struct reference<TypeIn, TypeOut, 1> {
+    static void forward_dft(const std::vector<std::int64_t> &sizes, TypeIn const *const in,
+                            TypeOut *out) {
+        reference_forward_dft_impl(in, out, sizes[0], 1);
+    }
+};
+
+template <typename TypeIn, typename TypeOut>
+struct reference<TypeIn, TypeOut, 2> {
+    static void forward_dft(const std::vector<std::int64_t> &sizes, TypeIn const *const in,
+                            TypeOut *out) {
+        const auto elements = std::accumulate(sizes.begin(), sizes.end(), 1, std::multiplies<>{});
+        std::vector<std::complex<double>> tmp(elements);
```

Comment:
Shouldn't we use `long double` for the temporary buffers here and below?

Reply:
addressed in https://github.com/oneapi-src/oneMKL/pull/275/commits/41f72310ccc0e6e81661d12cf658d0b8f7c3243a